src/native/http_request_response.c

```diff
@@ -332,60 +356,63 @@ static jobject s_make_request_general(
+        aws_http_stream_binding_release(env, stream_binding);
         return (jobject)NULL;
     }
 
     struct aws_http_make_request_options request_options = {
         .self_size = sizeof(request_options),
-        .request = callback_data->native_request,
+        .request = stream_binding->native_request,
         /* Set Callbacks */
         .on_response_headers = aws_java_http_stream_on_incoming_headers_fn,
         .on_response_header_block_done = aws_java_http_stream_on_incoming_header_block_done_fn,
         .on_response_body = aws_java_http_stream_on_incoming_body_fn,
         .on_complete = aws_java_http_stream_on_stream_complete_fn,
-        .user_data = callback_data,
+        .on_destroy = aws_java_http_stream_on_stream_destroy_fn,
+        .user_data = stream_binding,
     };
 
     jobject jHttpStreamBase = NULL;
 
-    callback_data->native_stream = aws_http_connection_make_request(native_conn, &request_options);
-    if (callback_data->native_stream) {
+    stream_binding->native_stream = aws_http_connection_make_request(native_conn, &request_options);
+    if (stream_binding->native_stream) {
         AWS_LOGF_TRACE(
             AWS_LS_HTTP_CONNECTION,
             "Opened new Stream on Connection. conn: %p, stream: %p",
             (void *)native_conn,
-            (void *)callback_data->native_stream);
+            (void *)stream_binding->native_stream);
 
-        jHttpStreamBase = aws_java_http_stream_from_native_new(env, callback_data, version);
+        jHttpStreamBase = aws_java_http_stream_from_native_new(env, stream_binding, version);
     }
 
     /* Check for errors that might have occurred while holding the lock. */
-    if (!callback_data->native_stream) {
-        /* Failed to create native aws_http_stream. Clean up callback_data. */
+    if (!stream_binding->native_stream) {
+        /* Failed to create native aws_http_stream. Clean up stream_binding. */
         AWS_LOGF_ERROR(AWS_LS_HTTP_CONNECTION, "Stream Request Failed. conn: %p", (void *)native_conn);
         aws_jni_throw_runtime_exception(env, "HttpClientConnection.MakeRequest: Unable to Execute Request");
-        aws_http_stream_binding_destroy(env, callback_data);
+        aws_http_stream_binding_release(env, stream_binding);
         return NULL;
     } else if (!jHttpStreamBase) {
-        /* Failed to create java HttpStream, but did create native aws_http_stream.
-          Close connection and mark native_stream for release.
-          callback_data will clean itself up when stream completes. */
-        aws_http_connection_close(native_conn);
-        aws_http_stream_release(callback_data->native_stream);
+        /* Failed to create java HttpStream, but did create native aws_http_stream. As we never activate the native
+         * stream, we can just release the native stream, and release the binding. */
+        aws_http_stream_release(stream_binding->native_stream);
+        aws_http_stream_binding_release(env, stream_binding);
```

Comment:
don't think this is right anymore.
before we run these 2 lines:
- stream_binding has refcount 1
- native_stream has refcount 1

then we call `aws_http_stream_release(stream_binding->native_stream);`
- stream_binding refcount goes to 0
    -  destroy callback fires...
    - native_stream refcount goes to 0

then we call `aws_http_stream_binding_release(env, stream_binding);`
kablooie, the stream_binding was already cleaned up

Reply:
the error handling in this function ... is just so confusing

man, we've gotten better at error handling over the years

this is bad old days stuff

Reply:
I pushed a cleanup of this function. I was one of the early authors. I feel some responsibility for this mess